### PR TITLE
dash/datagrid-mobile-responsiveness

### PR DESCRIPTION
**Pull request by Vazonik**
Added updating the data grid while container resizing, fixed overlapping headers and some minor bugs.

### CODE REVIEW

File: ts/DataGrid/DataGrid.ts [hunk 1]
```diff
@@ -514,6 +520,10 @@ class DataGrid {
         this.container.addEventListener('mouseover', (e): void => {
             this.handleMouseOver(e);
         });
+
+        new ResizeObserver((): void => {
```
Comment (goransle) [inline ts/DataGrid/DataGrid.ts:524]: Should we add this as a class prop, so DataGridComponent can call `observer.disconnect` on `destroy()`? (Possibly we should also have destroy on the DataGrid, but that's a larger matter)

File: ts/DataGrid/DataGrid.ts [hunk 2]
```diff
@@ -923,7 +946,8 @@ class DataGrid {
         }
 
         const headerEl = makeDiv(className);
-        headerEl.style.height = this.options.cellHeight + 'px';
+        headerEl.style.minHeight = this.options.cellHeight + 'px';
+        headerEl.style.maxHeight = this.options.cellHeight * 2 + 'px';
```
Comment (karolkolodziej) [inline ts/DataGrid/DataGrid.ts:950]: Please add the description in the `cellHeight` API that the header height might increase its size in certain cases.

File: ts/DataGrid/DataGrid.ts [hunk 3]
```diff
@@ -956,6 +980,89 @@ class DataGrid {
...
+
+        this.updateColumnHeaders();
+    }
+
+
+    /**
+     * Updates the column headers of the table.
+     * @internal
+     */
+    private updateColumnHeaders(): void {
```
Comment (karolkolodziej) [inline ts/DataGrid/DataGrid.ts:992]: I'm not sure if this only updated the header. I would change the name and would try to split it into smaller methods responsible for a given element.

**Review by karolkolodziej — CHANGES_REQUESTED** [review]

File: ts/DataGrid/DataGrid.ts (hunk 3 above)
Comment (bre1470) [reply ts/DataGrid/DataGrid.ts:992]: BTW Do we have any plans for row headers?

**Review by karolkolodziej — APPROVED** [review]
Looks good I'm only wondering why we don't get any visual changes. 🤔

File: test/cypress/integration/DataGrid/cell-header-format.cy.js [hunk 4]
```diff
@@ -53,11 +53,21 @@ describe('Remove the dashboard.', () => {
...
                 assert.closeTo(
                     finalWidth,
-                    500,
+                    200,
                     10,
                     'The width should be close to 500px.'
                 )
             });
+
+            cy.wait(50);
```
Comment (sebastianbochan) [inline test/cypress/integration/DataGrid/cell-header-format.cy.js:62]: What is the reason of `cy.wait` here?

File: ts/DataGrid/DataGrid.ts [hunk 5]
```diff
@@ -956,6 +979,115 @@ class DataGrid {
...
+        const headersContainer = this.columnHeadersContainer;
+        if (!headersContainer) {
+            return;
+        }
+
+        // Handle overflowing text in headers.
+        this.columnNames.forEach((columnName, i): void => {
```
Comment (pawellysy) [inline ts/DataGrid/DataGrid.ts:1008]: We avoid using `forEach` in the dashboards for performance reasons. Could you use the basic `for` loop?

**Review by pawellysy — CHANGES_REQUESTED** [review]

**Review by sebastianbochan — APPROVED** [review]
Please fix Pueblo suggestions.

**Comment by bre1470** [comment]
@goransle Is it possible to round the benchmark results to two fraction digits?

**Review by bre1470 — APPROVED** [review]
Nice!

**Review by pawellysy — APPROVED** [review]

**Pull request merged**